### PR TITLE
Service/Route Metrics

### PR DESCRIPTION
**Pull request by dliberman**
Add ability to collect metrics per service/route, maintaining
compatibility with previous version (only by service). This would
solve issue #40.

### CODE REVIEW

**Comment by dliberman** [comment]
This Travis CI build failed for something unrelated to the code changes, can someone verify if I made any mistake?

From [Jobs 361.1 and 361.2](https://travis-ci.com/Kong/kong-plugin-prometheus/jobs/190417498):
```
The command "git clone https://$GITHUB_TOKEN:@github.com/Kong/kong-ci.git" failed and exited with 128 during .
```

**Comment by hbagdi** [comment]
Hello @dliberman,

This has indeed been a request feature and it's time we add it.
Although, this will further introduce a performance penalty.

I think we should avoid introducing any more additional counters as that will further drive the contention and increase CPU load as reported in #43.


Could you change this to not have a config and the plugin by default, adds another label `route` label to all existing metrics which have a `service` label?

**Comment by dliberman** [comment]
Hi @hbagdi!

> I think we should avoid introducing any more additional counters as that will further drive the contention and increase CPU load as reported in #43.

Totally understand, but the way the code is proposed we'll be collecting by service **or** by service/route, so the number of counters on each request should remain the same.

> Could you change this to not have a config and the plugin by default, adds another label `route` label to all existing metrics which have a `service` label?

This is how we planned to do in the first place, as it totally works for us here :)
But, I was just thinking on remaining compatible with current plugin deploys. If the metrics generated suddenly get split also by routes, won't it break current dashboards, for instance? I believe those who still want to use only by service will have to aggregate/sum the stats per service to keep viewing the same information (so it will be a breaking change). Not sure if I'm right here.
But as a simple answer: if you believe it's better I can totally change the code to eliminate the config and only collect by service/route by default.

**Comment by hbagdi** [comment]
>  But, I was just thinking on remaining compatible with current plugin deploys. If the metrics generated suddenly get split also by routes, won't it break current dashboards, for instance? I believe those who still want to use only by service will have to aggregate/sum the stats per service to keep viewing the same information (so it will be a breaking change). Not sure if I'm right here.

@dliberman Good point, I've given this some thought:
All metrics in our [official Grafana Dashboard](https://grafana.com/dashboards/7424) use a `sum by` query, meaning if a `route` label is added, it will still do the correct aggregation, since the route label will be ignored. Even for alerting queries that users have written, most users should be using a `sum by` based query. While it is true that in some edge cases, this might not be the case and it might lead to breakage, making this change is worthwhile
- Having a separate timeseries in Prometheus for routes and services goes against the recommend practices. It will limit users to slice and dice the metrics in any way they want
- Since, the plugin currently doesn't support dynamic metrics, we should skip adding the config for now. I do want to add the capability to make metrics dynamic so that users can collect only metrics they really care about. Adding the config as proposed in this PR will make it hard in future to actually keep supporting it.


I'd suggest we stick with the simpler solution for now (add a route label to all metrics) till more capability is added to the plugin for dynamic labels and metrics.

**Comment by dliberman** [comment]
Hi @hbagdi.

> I'd suggest we stick with the simpler solution for now (add a route label to all metrics) till more capability is added to the plugin for dynamic labels and metrics.

I see your points and it's totally clear for me now, thanks!
I'll make the change and re-submit the PR.

File: kong/plugins/prometheus/exporter.lua [hunk 1]
```diff
@@ -54,31 +54,43 @@ local function log(message)
...
   end
 
-  metrics.status:inc(1, { message.response.status, service_name })
+  local route_name
+  if message and message.route then
+    if message.route.name then
+      route_name = message.route.name
+    else
+      route_name = message.route.id
+    end
```
Comment (hbagdi) [inline kong/plugins/prometheus/exporter.lua:63]: We can simplify this logic as:
```lua
route_name = message.route.name or message.route.id
```

This is a handy way in Lua to to route_name to id when name is nil.

**Review by hbagdi — CHANGES_REQUESTED** [review]
Thank you @dliberman 

Please add/update tests, otherwise this is looking good.

**Comment by jdomag** [comment]
@dliberman 
Are you still working on this? Would be great to have a per route metrics :)

**Comment by xichengliudui** [comment]
This is a very good idea. Can we proceed with it?

**Comment by wpjunior** [comment]
Nice patch!, It is so important when you need to separate service metrics by kind of criticality,  @dliberman and @hbagdi, could you proceed with MR?

**Comment by hbagdi** [comment]
Update:
I've cleaned this patch and will be merging in next week with tests!

**Comment by dliberman** [comment]
Thank  you @hbagdi!
I've been working on several projects other than Kong for a few months already, so I wasn't in touch with it. The patch looks good, was just missing some tests to be submitted.

**Comment by hbagdi** [comment]
Merged in https://github.com/Kong/kong-plugin-prometheus/commit/8bca59408a3c1c83d5a1a3742fd566eadb7ef70a

**Comment by xichengliudui** [comment]
How do I use it? I don't use lua myself?

**Comment by hbagdi** [comment]
This will ship with Kong 1.4.
Please wait for the release.

You can also use `luarocks` to upgrade the Prometheus plugin that is installed alongside Kong.

**Pull request closed without merging**